### PR TITLE
Connection cache

### LINKED ISSUE #24: Connection pooling

### PR DESCRIPTION
**Pull request by SeriyBg**
Reusing existing client instances for the same URL
Fix #24

### CODE REVIEW

File: src/test/java/com/hazelcast/jdbc/JdbcPreparedStatementTest.java [hunk 1]
```diff
@@ -34,25 +34,28 @@
...
         IMap<Integer, Person> personMap = member.getMap("person");
         for (int i = 0; i < 3; i++) {
             personMap.put(i, new Person("Jack"+i, i));
         }
     }
 
-    @AfterEach
-    public void tearDown() {
+    @AfterAll
+    public static void tearDown() {
```
Comment (hakanmemisoglu) [inline src/test/java/com/hazelcast/jdbc/JdbcPreparedStatementTest.java:52]: Do we need a `connection.close()` statement here?

**Review by alex-dukhno — APPROVED** [review]

File: src/main/java/com/hazelcast/jdbc/JdbcUrl.java [hunk 2]
```diff
@@ -64,6 +65,23 @@ public String getRawAuthority() {
...
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        JdbcUrl jdbcUrl = (JdbcUrl) o;
+        return Objects.equals(properties, jdbcUrl.properties) && Objects.equals(rawUrl, jdbcUrl.rawUrl);
```
Comment (viliam-durina) [inline src/main/java/com/hazelcast/jdbc/JdbcUrl.java:77]: I think it's enough to compare the `rawUrl` - it includes the properties.

File: src/main/java/com/hazelcast/jdbc/HazelcastSqlClient.java [hunk 3]
```diff
@@ -39,6 +49,7 @@ SqlResult execute(SqlStatement sqlStatement) {
 
     void shutdown() {
         client.shutdown();
```
Comment (viliam-durina) [inline src/main/java/com/hazelcast/jdbc/HazelcastSqlClient.java:51]: We close the client with first closed connection, but there can be multiple connections using that same client.

The following test if added to `DriverIMDGTest` fails:
```java
    @Test
    void when_twoConnectionsCreatedAndOneClosed_then_theClientShouldRemain() throws SQLException {
        Connection connection2 = DriverManager.getConnection(JDBC_HAZELCAST_LOCALHOST);
        connection2.close();
        connection.createStatement().execute("SELECT * FROM person");
    }
```
It throws:
```
com.hazelcast.client.HazelcastClientNotActiveException: Client is not active.

	at com.hazelcast.client.impl.clientside.HazelcastClientProxy.getClient(HazelcastClientProxy.java:303)
	at com.hazelcast.client.impl.clientside.HazelcastClientProxy.getSql(HazelcastClientProxy.java:319)
	at com.hazelcast.jdbc.HazelcastSqlClient.execute(HazelcastSqlClient.java:47)
```
We need to do some reference counting.

**Review by viliam-durina — CHANGES_REQUESTED** [review]

File: src/main/java/com/hazelcast/jdbc/JdbcUrl.java (hunk 2 above)
Comment (SeriyBg) [reply src/main/java/com/hazelcast/jdbc/JdbcUrl.java:77]: Not always, as the properties might be passed using of the methods `DriverManager.getConnection(String URL, java.util.Properties info)` or `DriverManager.getConnection(String URL, String user, String password)`

**Comment by viliam-durina** [comment]
There's one more issue with the cached connections: we have to close all results that were created using that connection. E.g.:
```
Connection conn1 = DriverManager.getConnection(...);
Connection conn2 = DriverManager.getConnection(...);
conn1.createStatement().execute(...);
conn2.createStatement().execute(...);
conn1.close();
```
The statement created using `conn1` must be closed (cancelled on server), but the stmt crated using `conn2` must continue.

**Comment by SeriyBg** [comment]
With the issue mentioned by @viliam-durina it seems that connection caching is an error-prone approach that will unnecessarily complicate the implementation. Thus, for now, I'll close the PR

**Pull request closed without merging**